File: net/FlatBuffers/ByteBuffer.cs
/*
 * Copyright 2014 Google Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// There are 2 #defines that have an impact on performance of this ByteBuffer implementation
//
//      UNSAFE_BYTEBUFFER 
//          This will use unsafe code to manipulate the underlying byte array. This
//          can yield a reasonable performance increase.
//
//      BYTEBUFFER_NO_BOUNDS_CHECK
//          This will disable the bounds check asserts to the byte array. This can
//          yield a small performance gain in normal code..
//
// Using UNSAFE_BYTEBUFFER and BYTEBUFFER_NO_BOUNDS_CHECK together can yield a
// performance gain of ~15% for some operations, however doing so is potentially 
// dangerous. Do so at your own risk!
//

using System;
using System.IO;
using System.Text;

namespace FlatBuffers
{
    /// <summary>
    /// Class to mimic Java's ByteBuffer which is used heavily in Flatbuffers.
    /// </summary>
    public class ByteBuffer
    {
        protected byte[] _buffer;
        private int _pos;  // Must track start of the buffer.

        public int Length { get { return _buffer.Length; } }

        public ByteBuffer(int size) : this(new byte[size]) { }

        public ByteBuffer(byte[] buffer) : this(buffer, 0) { }

        public ByteBuffer(byte[] buffer, int pos)
        {
            _buffer = buffer;
            _pos = pos;
        }

        public int Position {
            get { return _pos; }
            set { _pos = value; }
        }

        public void Reset()
        {
            _pos = 0;
        }

        // Create a new ByteBuffer on the same underlying data.
        // The new ByteBuffer's position will be same as this buffer's.
        public ByteBuffer Duplicate()
        {
            return new ByteBuffer(_buffer, Position);
        }

        // Increases the size of the ByteBuffer, and copies the old data towards
        // the end of the new buffer.
        public void GrowFront(int newSize)
        {
            if ((Length & 0xC0000000) != 0)
                throw new Exception(
                    "ByteBuffer: cannot grow buffer beyond 2 gigabytes.");

            if (newSize < Length)
                throw new Exception("ByteBuffer: cannot truncate buffer.");

            byte[] newBuffer = new byte[newSize];
            Buffer.BlockCopy(_buffer, 0, newBuffer, newSize - Length,
                             Length);
            _buffer = newBuffer;
        }

        public byte[] ToArray(int pos, int len)
        {
            byte[] arr = new byte[len];
            Buffer.BlockCopy(_buffer, pos, arr, 0, len);
            return arr;
        }

        public byte[] ToSizedArray()
        {
            return ToArray(Position, Length - Position);
        }

        public byte[] ToFullArray()
        {
            return ToArray(0, Length);
        }

        public ArraySegment<byte> ToArraySegment(int pos, int len)
        {
            return new ArraySegment<byte>(_buffer, pos, len);
        }

        public MemoryStream ToMemoryStream(int pos, int len)
        {
            return new MemoryStream(_buffer, pos, len);
        }

#if !UNSAFE_BYTEBUFFER
        // Pre-allocated helper arrays for convertion.
        private float[] floathelper = new[] { 0.0f };
        private int[] inthelper = new[] { 0 };
        private double[] doublehelper = new[] { 0.0 };
        private ulong[] ulonghelper = new[] { 0UL };
#endif // !UNSAFE_BYTEBUFFER

        // Helper functions for the unsafe version.
        static public ushort ReverseBytes(ushort input)
        {
            return (ushort)(((input & 0x00FFU) << 8) |
                            ((input & 0xFF00U) >> 8));
        }
        static public uint ReverseBytes(uint input)
        {
            return ((input & 0x000000FFU) << 24) |
                   ((input & 0x0000FF00U) <<  8) |
                   ((input & 0x00FF0000U) >>  8) |
                   ((input & 0xFF000000U) >> 24);
        }
        static public ulong ReverseBytes(ulong input)
        {
            return (((input & 0x00000000000000FFUL) << 56) |
                    ((input & 0x000000000000FF00UL) << 40) |
                    ((input & 0x0000000000FF0000UL) << 24) |
                    ((input & 0x00000000FF000000UL) <<  8) |
                    ((input & 0x000000FF00000000UL) >>  8) |
                    ((input & 0x0000FF0000000000UL) >> 24) |
                    ((input & 0x00FF000000000000UL) >> 40) |
                    ((input & 0xFF00000000000000UL) >> 56));
        }

#if !UNSAFE_BYTEBUFFER
        // Helper functions for the safe (but slower) version.
        protected void WriteLittleEndian(int offset, int count, ulong data)
        {
            if (BitConverter.IsLittleEndian)
            {
                for (int i = 0; i < count; i++)
                {
                    _buffer[offset + i] = (byte)(data >> i * 8);
                }
            }
            else
            {
                for (int i = 0; i < count; i++)
                {
                    _buffer[offset + count - 1 - i] = (byte)(data >> i * 8);
                }
            }
        }

        protected ulong ReadLittleEndian(int offset, int count)
        {
            AssertOffsetAndLength(offset, count);
            ulong r = 0;
            if (BitConverter.IsLittleEndian)
            {
                for (int i = 0; i < count; i++)
                {
                  r |= (ulong)_buffer[offset + i] << i * 8;
                }
            }
            else
            {
              for (int i = 0; i < count; i++)
              {
                r |= (ulong)_buffer[offset + count - 1 - i] << i * 8;
              }
            }
            return r;
        }
#endif // !UNSAFE_BYTEBUFFER

        private void AssertOffsetAndLength(int offset, int length)
        {
            #if !BYTEBUFFER_NO_BOUNDS_CHECK
            if (offset < 0 ||
                offset > _buffer.Length - length)
                throw new ArgumentOutOfRangeException();
            #endif
        }

        public void PutSbyte(int offset, sbyte value)
        {
            AssertOffsetAndLength(offset, sizeof(sbyte));
            _buffer[offset] = (byte)value;
        }

        public void PutByte(int offset, byte value)
        {
            AssertOffsetAndLength(offset, sizeof(byte));
            _buffer[offset] = value;
        }

        public void PutByte(int offset, byte value, int count)
        {
            AssertOffsetAndLength(offset, sizeof(byte) * count);
            for (var i = 0; i < count; ++i)
                _buffer[offset + i] = value;
        }

        // this method exists in order to conform with Java ByteBuffer standards
        public void Put(int offset, byte value)
        {
            PutByte(offset, value);
        }

        public void PutStringUTF8(int offset, string value)
        {
            AssertOffsetAndLength(offset, value.Length);
            Encoding.UTF8.GetBytes(value, 0, value.Length,
                _buffer, offset);
        }

#if UNSAFE_BYTEBUFFER
        // Unsafe but more efficient versions of Put*.
        public void PutShort(int offset, short value)
        {
            PutUshort(offset, (ushort)value);
        }

        public unsafe void PutUshort(int offset, ushort value)
        {
            AssertOffsetAndLength(offset, sizeof(ushort));
            fixed (byte* ptr = _buffer)
            {
                *(ushort*)(ptr + offset) = BitConverter.IsLittleEndian
                    ? value
                    : ReverseBytes(value);
            }
        }

        public void PutInt(int offset, int value)
        {
            PutUint(offset, (uint)value);
        }

        public unsafe void PutUint(int offset, uint value)
        {
            AssertOffsetAndLength(offset, sizeof(uint));
            fixed (byte* ptr = _buffer)
            {
                *(uint*)(ptr + offset) = BitConverter.IsLittleEndian
                    ? value
                    : ReverseBytes(value);
            }
        }

        public unsafe void PutLong(int offset, long value)
        {
            PutUlong(offset, (ulong)value);
        }

        public unsafe void PutUlong(int offset, ulong value)
        {
            AssertOffsetAndLength(offset, sizeof(ulong));
            fixed (byte* ptr = _buffer)
            {
                *(ulong*)(ptr + offset) = BitConverter.IsLittleEndian
                    ? value
                    : ReverseBytes(value);
            }
        }

        public unsafe void PutFloat(int offset, float value)
        {
            AssertOffsetAndLength(offset, sizeof(float));
            fixed (byte* ptr = _buffer)
            {
                if (BitConverter.IsLittleEndian)
                {
                    *(float*)(ptr + offset) = value;
                }
                else
                {
                    *(uint*)(ptr + offset) = ReverseBytes(*(uint*)(&value));
                }
            }
        }

        public unsafe void PutDouble(int offset, double value)
        {
            AssertOffsetAndLength(offset, sizeof(double));
            fixed (byte* ptr = _buffer)
            {
                if (BitConverter.IsLittleEndian)
                {
                    *(double*)(ptr + offset) = value;

                }
                else
                {
                    *(ulong*)(ptr + offset) = ReverseBytes(*(ulong*)(ptr + offset));
                }
            }
        }
#else // !UNSAFE_BYTEBUFFER
        // Slower versions of Put* for when unsafe code is not allowed.
        public void PutShort(int offset, short value)
        {
            AssertOffsetAndLength(offset, sizeof(short));
            WriteLittleEndian(offset, sizeof(short), (ulong)value);
        }

        public void PutUshort(int offset, ushort value)
        {
            AssertOffsetAndLength(offset, sizeof(ushort));
            WriteLittleEndian(offset, sizeof(ushort), (ulong)value);
        }

        public void PutInt(int offset, int value)
        {
            AssertOffsetAndLength(offset, sizeof(int));
            WriteLittleEndian(offset, sizeof(int), (ulong)value);
        }

        public void PutUint(int offset, uint value)
        {
            AssertOffsetAndLength(offset, sizeof(uint));
            WriteLittleEndian(offset, sizeof(uint), (ulong)value);
        }

        public void PutLong(int offset, long value)
        {
            AssertOffsetAndLength(offset, sizeof(long));
            WriteLittleEndian(offset, sizeof(long), (ulong)value);
        }

        public void PutUlong(int offset, ulong value)
        {
            AssertOffsetAndLength(offset, sizeof(ulong));
            WriteLittleEndian(offset, sizeof(ulong), value);
        }

        public void PutFloat(int offset, float value)
        {
            AssertOffsetAndLength(offset, sizeof(float));
            floathelper[0] = value;
            Buffer.BlockCopy(floathelper, 0, inthelper, 0, sizeof(float));
            WriteLittleEndian(offset, sizeof(float), (ulong)inthelper[0]);
        }

        public void PutDouble(int offset, double value)
        {
            AssertOffsetAndLength(offset, sizeof(double));
            doublehelper[0] = value;
            Buffer.BlockCopy(doublehelper, 0, ulonghelper, 0, sizeof(double));
            WriteLittleEndian(offset, sizeof(double), ulonghelper[0]);
        }

#endif // UNSAFE_BYTEBUFFER

        public sbyte GetSbyte(int index)
        {
            AssertOffsetAndLength(index, sizeof(sbyte));
            return (sbyte)_buffer[index];
        }

        public byte Get(int index)
        {
            AssertOffsetAndLength(index, sizeof(byte));
            return _buffer[index];
        }

        public string GetStringUTF8(int startPos, int len)
        {
            return Encoding.UTF8.GetString(_buffer, startPos, len);
        }

#if UNSAFE_BYTEBUFFER
        // Unsafe but more efficient versions of Get*.
        public short GetShort(int offset)
        {
            return (short)GetUshort(offset);
        }

        public unsafe ushort GetUshort(int offset)
        {
            AssertOffsetAndLength(offset, sizeof(ushort));
            fixed (byte* ptr = _buffer)
            {
                return BitConverter.IsLittleEndian
                    ? *(ushort*)(ptr + offset)
                    : ReverseBytes(*(ushort*)(ptr + offset));
            }
        }

        public int GetInt(int offset)
        {
            return (int)GetUint(offset);
        }

        public unsafe uint GetUint(int offset)
        {
            AssertOffsetAndLength(offset, sizeof(uint));
            fixed (byte* ptr = _buffer)
            {
                return BitConverter.IsLittleEndian
                    ? *(uint*)(ptr + offset)
                    : ReverseBytes(*(uint*)(ptr + offset));
            }
        }

        public long GetLong(int offset)
        {
            return (long)GetUlong(offset);
        }

        public unsafe ulong GetUlong(int offset)
        {
            AssertOffsetAndLength(offset, sizeof(ulong));
            fixed (byte* ptr = _buffer)
            {
                return BitConverter.IsLittleEndian
                    ? *(ulong*)(ptr + offset)
                    : ReverseBytes(*(ulong*)(ptr + offset));
            }
        }

        public unsafe float GetFloat(int offset)
        {
            AssertOffsetAndLength(offset, sizeof(float));
            fixed (byte* ptr = _buffer)
            {
                if (BitConverter.IsLittleEndian)
                {
                    return *(float*)(ptr + offset);
                }
                else
                {
                    uint uvalue = ReverseBytes(*(uint*)(ptr + offset));
                    return *(float*)(&uvalue);
                }
            }
        }

        public unsafe double GetDouble(int offset)
        {
            AssertOffsetAndLength(offset, sizeof(double));
            fixed (byte* ptr = _buffer)
            {
                if (BitConverter.IsLittleEndian)
                {
                    return *(double*)(ptr + offset);
                }
                else
                {
                    ulong uvalue = ReverseBytes(*(ulong*)(ptr + offset));
                    return *(double*)(&uvalue);
                }
            }
        }
#else // !UNSAFE_BYTEBUFFER
        // Slower versions of Get* for when unsafe code is not allowed.
        public short GetShort(int index)
        {
            return (short)ReadLittleEndian(index, sizeof(short));
        }

        public ushort GetUshort(int index)
        {
            return (ushort)ReadLittleEndian(index, sizeof(ushort));
        }

        public int GetInt(int index)
        {
            return (int)ReadLittleEndian(index, sizeof(int));
        }

        public uint GetUint(int index)
        {
            return (uint)ReadLittleEndian(index, sizeof(uint));
        }

        public long GetLong(int index)
        {
           return (long)ReadLittleEndian(index, sizeof(long));
        }

        public ulong GetUlong(int index)
        {
            return ReadLittleEndian(index, sizeof(ulong));
        }

        public float GetFloat(int index)
        {
            int i = (int)ReadLittleEndian(index, sizeof(float));
            inthelper[0] = i;
            Buffer.BlockCopy(inthelper, 0, floathelper, 0, sizeof(float));
            return floathelper[0];
        }

        public double GetDouble(int index)
        {
            ulong i = ReadLittleEndian(index, sizeof(double));
            // There's Int64BitsToDouble but it uses unsafe code internally.
            ulonghelper[0] = i;
            Buffer.BlockCopy(ulonghelper, 0, doublehelper, 0, sizeof(double));
            return doublehelper[0];
        }
#endif // UNSAFE_BYTEBUFFER
    }
}

File: php/ByteBuffer.php
<?php

/*
 * Copyright 2015 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

declare(strict_types=1);

namespace Google\FlatBuffers;

use Exception;

// phpcs:disable PSR2.Classes.PropertyDeclaration.Underscore

// /**
//  * @return bool
//  */
// function is_little_endian(): bool
// {
//     $v = unpack('S', "\x01\x00");
//     assert($v !== false && array_key_exists(1, $v) && is_int($v[1]));
//     return $v[1] === 1;
// }

// if (!is_little_endian()) {
//     throw new Exception('Big-endian platform is not supported.');
// }

/**
 * @phpstan-type SizeOfType Constants::SIZEOF_BOOL|Constants::SIZEOF_BYTE|Constants::SIZEOF_SHORT|Constants::SIZEOF_INT|Constants::SIZEOF_LONG|Constants::SIZEOF_FLOAT|Constants::SIZEOF_DOUBLE|Constants::FILE_IDENTIFIER_LENGTH
 */
class ByteBuffer
{
    /**
     * @var string $_buffer;
     */
    public string $_buffer;

    /**
     * @var NPosT $npos;
     */
    private int $npos = 0;

    /**
     * @param string $bytes
     * @return ByteBuffer
     */
    public static function wrap(string $bytes): ByteBuffer
    {
        $bb = new ByteBuffer(0);
        $bb->_buffer = $bytes;

        return $bb;
    }

    /**
     * @param BufSizeT $size
     */
    public function __construct(int $size)
    {
        $this->_buffer = str_repeat("\0", $size);
    }

    /**
     * @return ByteBuffer
     * @throws \Exception Exception.
     */
    public function createByGrowing(): self
    {
        $old_size = $this->capacity();
        if (($old_size & 0xC0000000) !== 0) {
            throw new \Exception("FlatBuffers: cannot grow buffer beyond 2 gigabytes");
        }

        // Double the size of buffer.
        $bb = new self($old_size);
        $bb->_buffer .= $this->_buffer;

        $bb->setPosition($old_size); // Is this correct?

        return $bb;
    }

    /**
     * @return BufSizeT
     */
    public function capacity(): int
    {
        return Constants::asBufSize(strlen($this->_buffer));
    }

    /**
     * @return NPosT
     */
    public function getPosition(): int
    {
        return $this->npos;
    }

    /**
     * @param NPosT $npos
     */
    public function setPosition(int $npos): void
    {
        $this->npos = $npos;
    }

    /**
     *
     */
    public function reset(): void
    {
        $this->npos = 0;
    }

    /**
     * @return BufSizeT
     */
    public function length(): int
    {
        return Constants::asBufSize(strlen($this->_buffer) - $this->npos);
    }

    /**
     * @return string
     */
    public function data(): string
    {
        return substr($this->_buffer, $this->npos);
    }

    /**
     * @param NPosT    $npos
     * @param BufSizeT $size
     */
    public function assertOffsetAndLength(int $npos, int $size): void
    {
        if (strlen($this->_buffer) < $npos + $size) {
            throw new \OutOfRangeException(
                sprintf("npos: %d, size: %d, buffer; %d", $npos, $size, strlen($this->_buffer))
            );
        }
    }

    /**
     * @param NPosT            $npos
     * @param SizeOfType       $size
     * @param string           $pack_code
     * @param int|float|double $value
     */
    private function writeToBuffer(int $npos, int $size, string $pack_code, int|float $value): void
    {
        $this->assertOffsetAndLength($npos, $size);
        $bytes = pack($pack_code, $value);
        assert(strlen($bytes) === $size);
        $this->_buffer = substr_replace($this->_buffer, $bytes, $npos, $size);
    }

    /**
     * @param NPosT      $npos
     * @param SizeOfType $size
     * @param string     $pack_code
     * @return int|float|double
     */
    private function readFromBuffer(int $npos, int $size, string $pack_code): int|float
    {
        $this->assertOffsetAndLength($npos, $size);
        $v = unpack($pack_code, $this->_buffer, $npos);
        assert($v !== false && array_key_exists(1, $v));
        return $v[1];
    }

    /**
     * @param NPosT          $npos
     * @param int<-128, 127> $value
     */
    public function putSbyte(int $npos, int $value): void
    {
        self::validateInt(-128, 127, $value, "sbyte");
        $this->writeToBuffer($npos, Constants::SIZEOF_BYTE, 'c', $value);
    }

    /**
     * @param NPosT       $npos
     * @param int<0, 255> $value
     */
    public function putByte(int $npos, int $value): void
    {
        self::validateInt(0, 255, $value, "byte");
        $this->writeToBuffer($npos, Constants::SIZEOF_BYTE, 'C', $value);
    }

    /**
     * @param NPosT  $npos
     * @param string $value
     */
    public function put(int $npos, string $value): void
    {
        $size = Constants::asBufSize(strlen($value));
        $this->assertOffsetAndLength($npos, $size);
        $this->_buffer = substr_replace($this->_buffer, $value, $npos, $size);
    }

    /**
     * @param NPosT $npos
     * @param bool  $value
     */
    public function putBool(int $npos, bool $value): void
    {
        $this->writeToBuffer($npos, Constants::SIZEOF_BYTE, 'C', (int)$value);
    }

    /**
     * @param NPosT              $npos
     * @param int<-32768, 32767> $value
     */
    public function putShort(int $npos, int $value): void
    {
        self::validateInt(-32768, 32767, $value, "short");
        $this->writeToBuffer($npos, Constants::SIZEOF_SHORT, 's', $value);
    }

    /**
     * @param NPosT         $npos
     * @param int<0, 65535> $value
     */
    public function putUshort(int $npos, int $value): void
    {
        self::validateInt(0, 65535, $value, "ushort");
        $this->writeToBuffer($npos, Constants::SIZEOF_SHORT, 'v', $value);
    }

    /**
     * @param NPosT                        $npos
     * @param int<-2147483648, 2147483647> $value
     */
    public function putInt(int $npos, int $value): void
    {
        // 2147483647 = (1 << 31) -1 = Maximum signed 32-bit int
        // -2147483648 = -1 << 31 = Minimum signed 32-bit int
        self::validateInt(-2147483648, 2147483647, $value, "int");
        $this->writeToBuffer($npos, Constants::SIZEOF_INT, 'i', $value);
    }

    /**
     * @param NPosT    $npos
     * @param SOffsetT $value
     */
    public function putSOffset(int $npos, int $value): void
    {
        $this->putInt($npos, $value);
    }

    /**
     * @param NPosT              $npos
     * @param int<0, 4294967295> $value
     */
    public function putUint(int $npos, int $value): void
    {
        // NOTE: We can't put big integer value. this is PHP limitation.
        // 4294967295 = (1 << 32) -1 = Maximum unsigned 32-bin int
        self::validateInt(0, 4294967295, $value, "uint", " php has big numbers limitation. check your PHP_INT_MAX");
        $this->writeToBuffer($npos, Constants::SIZEOF_INT, 'I', $value);
    }

    /**
     * @param NPosT $npos
     * @param int   $value
     */
    public function putLong(int $npos, int $value): void
    {
        // NOTE: We can't put big integer value. this is PHP limitation.
        self::validateInt(
            ~PHP_INT_MAX,
            PHP_INT_MAX,
            $value,
            "long",
            " php has big numbers limitation. check your PHP_INT_MAX"
        );
        $this->writeToBuffer($npos, Constants::SIZEOF_LONG, 'q', $value);
    }

    /**
     * @param NPosT       $npos
     * @param int<0, max> $value
     */
    public function putUlong(int $npos, int $value): void
    {
        // NOTE: We can't put big integer value. this is PHP limitation.
        self::validateInt(0, PHP_INT_MAX, $value, "long", " php has big numbers limitation. check your PHP_INT_MAX");
        $this->writeToBuffer($npos, Constants::SIZEOF_LONG, 'P', $value);
    }

    /**
     * @param NPosT $npos
     * @param float $value
     */
    public function putFloat(int $npos, float $value): void
    {
        $this->assertOffsetAndLength($npos, Constants::SIZEOF_FLOAT);
        $this->writeToBuffer($npos, Constants::SIZEOF_FLOAT, 'f', $value);
    }

    /**
     * @param NPosT  $npos
     * @param double $value
     */
    public function putDouble(int $npos, float $value): void
    {
        $this->assertOffsetAndLength($npos, Constants::SIZEOF_DOUBLE);
        $this->writeToBuffer($npos, Constants::SIZEOF_DOUBLE, 'd', $value);
    }

    /**
     * @param NPosT $npos
     * @return int<0, 255>
     */
    public function getByte(int $npos): int
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_BYTE, 'C');
        assert(is_int($value) && 0 <= $value && $value <= 255);
        return $value;
    }

    /**
     * @param NPosT $npos
     * @return int<-128, 127>
     */
    public function getSbyte(int $npos): int
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_BYTE, 'c');
        assert(is_int($value) && -128 <= $value && $value <= 127);
        return $value;
    }

    /**
     * @param NPosT    $npos
     * @param BufSizeT $size
     * @return string
     */
    public function get(int $npos, int $size): string
    {
        $this->assertOffsetAndLength($npos, $size);
        return substr($this->_buffer, $npos, $size);
    }

    /**
     * @param NPosT $npos
     * @return bool
     */
    public function getBool(int $npos): bool
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_BYTE, 'C');
        assert(is_int($value) && 0 <= $value && $value <= 255);
        return (bool)$value;
    }

    /**
     * @param NPosT $npos
     * @return int<-32768, 32767>
     */
    public function getShort(int $npos): int
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_SHORT, 's');
        assert(is_int($value) && -32768 <= $value && $value <= 32767);
        return $value;
    }

    /**
     * @param NPosT $npos
     * @return int<0, 65535>
     */
    public function getUshort(int $npos): int
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_SHORT, 'v');
        assert(is_int($value) && 0 <= $value && $value <= 65535);
        return $value;
    }

    /**
     * @param NPosT $npos
     * @return VOffsetT
     */
    public function getVOffset(int $npos): int
    {
        return $this->getUshort($npos);
    }

    /**
     * @param NPosT $npos
     * @return int<-2147483648, 2147483647>
     */
    public function getInt(int $npos): int
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_INT, 'i');
        assert(is_int($value) && -2147483648 <= $value && $value <= 2147483647);
        return $value;
    }

    /**
     * @param NPosT $npos
     * @return SOffsetT
     */
    public function getSOffset(int $npos): int
    {
        return $this->getInt($npos);
    }

    /**
     * @param NPosT $npos
     * @return int<0, 4294967295>
     */
    public function getUint(int $npos): int
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_INT, 'I');
        assert(is_int($value) && 0 <= $value && $value <= 4294967295);
        return $value;
    }

    /**
     * @param NPosT $npos
     * @return UOffsetT
     */
    public function getUOffset(int $npos): int
    {
        return $this->getUint($npos);
    }

    /**
     * @param NPosT $npos
     * @return int
     */
    public function getLong(int $npos): int
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_LONG, 'q');
        assert(is_int($value));
        return $value;
    }

    /**
     * @param NPosT $npos
     * @return int<0, max>
     */
    public function getUlong(int $npos): int
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_LONG, 'P');
        assert(is_int($value) && 0 <= $value);
        return $value;
    }

    /**
     * @param NPosT $npos
     * @return float
     */
    public function getFloat(int $npos): float
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_FLOAT, 'f');
        assert(is_float($value));
        return $value;
    }

    /**
     * @param NPosT $npos
     * @return double
     */
    public function getDouble(int $npos): float
    {
        $value = $this->readFromBuffer($npos, Constants::SIZEOF_DOUBLE, 'd');
        assert(is_double($value));
        return $value;
    }

    /**
     * @param int    $min
     * @param int    $max
     * @param int    $value
     * @param string $type
     * @param string $additional_notes
     */
    private static function validateInt(
        int $min,
        int $max,
        int $value,
        string $type,
        string $additional_notes = ""
    ): void {
        if ($min <= $value && $value <= $max) {
            return;
        }

        throw new \InvalidArgumentException(sprintf("bad number %s for type %s.%s", $value, $type, $additional_notes));
    }

    /**
     * @param NPosT $t_npos
     * @return NPosT
     */
    public function followSOffset(int $t_npos): int
    {
        return Constants::asNPos($t_npos - $this->getSOffset($t_npos));
    }

    /**
     * @param NPosT $npos
     * @return NPosT
     */
    public function followUOffset(int $npos): int
    {
        return Constants::asNPos($this->getUOffset($npos) + $npos);
    }

    /**
     * Write position to normal position conversion.
     *
     * @param WPosT $wpos
     * @return NPosT
     */
    public function wposToNpos(int $wpos): int
    {
        return Constants::asNPos($this->capacity() - $wpos);
    }

    /**
     * Normal position to write position conversion.
     *
     * @param NPosT $npos
     * @return WPosT
     */
    public function nposToWpos(int $npos): int
    {
        return Constants::asWPos($this->capacity() - $npos);
    }
}
